### PR TITLE
feat: support filtering in trial logs api [DET-4177]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change adds filtering support to `/api/v1/trial/{trial_id}/logs`. It also unifies the trial and master logs streaming code paths and provides a small internal filter representation and a not-too-general way to convert it to SQL.

## Test Plan
- [x] add integration tests for the complex bits
- [x] test notebook, trial, master logs all still work as expected.

## Commentary (optional)
There were a lot of ways to do all this. For the most part, in the end, I settled back on the simpler way.

The only this in the PR I think now that is mildly controversial is the `api.ProcessLogs` is now an actor `LogStoreProcessor` or something. The idea here was mainly motivated by this opinion of mine: everything (especially things that are even remotely long-running, like streaming trial logs) should just be an actor. Now that some preliminary tracing support is going to land, it's even easier to get visibility into that plain-old goroutines, and the more things are actors, the more useful the actors become, imo.

## Checklist

- [x] No release notes until there are fields to actually filter on populated.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by stoksc** [comment]
This PR is based on (and blocked on) #1373.

File: master/internal/api/logs.go [hunk 1]
```diff
@@ -6,121 +6,166 @@ import (
...
 					return nil
 				}
 			}
-			time.Sleep(logCheckWaitTime)
-		}
-		if !req.Follow || req.Limit == 0 {
-			return nil
-		} else if req.Follow {
-			req.Limit = -1
+			actors.NotifyAfter(ctx, 5*time.Second, tick{})
```
Comment (stoksc) [inline master/internal/api/logs.go:102]: discuss

File: master/internal/db/postgres_filters.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	for _, f := range fs {
+		filterParams := filterToParams(f)
+		fragments = append(fragments, filterToSQL(f, filterParams, paramID))
+		params = append(params, filterParams...)
+		paramID += len(filterParams)
+	}
+	return strings.Join(fragments, "\n"), params
+}
+
+func filterToSQL(f api.Filter, values []interface{}, paramID int) string {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/db/postgres_filters.go:26]: At least in the current code path, this should only ever come from internal, but to be safe going forward I added this and a comment that says `api.Filter.Field` should _never_ be user input.

File: master/internal/api/logs.go [hunk 3]
```diff
@@ -6,121 +6,166 @@ import (
...
+		default:
+			l.req.Limit -= batch.Size()
+			l.req.Offset += batch.Size()
+			switch err := l.process(batch); {
+			case err != nil:
+				return fmt.Errorf("failed while processing batch: %w", err)
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/api/logs.go:108]: I think `fmt.Errorf` with the `%w` formatting directive is the "new" way to do this. I played with it in a playground and the behavior seems identical. https://stackoverflow.com/questions/61933650/whats-the-difference-between-errors-wrapf-errors-errorf-and-fmt-errorf

[the remark this answers is not in the document]
Comment (dzhu) [reply master/internal/api/logs.go:110]: Oh, interesting; I keep forgetting about the new error APIs. I'd still be mildly inclined to stick with the one that is currently used much more in our code (and I guess it has support for stack traces too, not that we use that anywhere), pending a later sweeping refactor. But it doesn't matter much.

**Comment by stoksc** [comment]
@dzhu oops, looks like I _unassigned_ you after I fixed this up last time 🤦

File: master/internal/api_trials.go [hunk 4]
```diff
@@ -42,53 +50,135 @@ func (a *apiServer) TrialLogs(
...
 
-	offset := api.EffectiveOffset(int(req.Offset), total)
+	filters, err := constructTrialLogsFilters(req)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, fmt.Sprintf("unsupported filter: %s", err))
+	}
 
-	if limit := int32(total - offset); !req.Follow && (limit < req.Limit || req.Limit == 0) {
-		req.Limit = limit
+	logID := int32(-1) // WebUI assumes logs are 0-indexed
```
Comment (dzhu) [inline master/internal/api_trials.go:68]:
```suggestion
	logID := int32(-1) // WebUI assumes logs are 0-indexed.
```
and on a separate line.

File: proto/src/determined/api/v1/trial.proto [hunk 5]
```diff
@@ -20,7 +26,24 @@ message TrialLogsRequest {
...
+  repeated string agent_ids = 6;
+  // Limit the trial logs to a subset of containers.
+  repeated string container_ids = 7;
+  // Limit the trial logs to a subset of ranks.
+  repeated int32 rank_ids = 8;
+  // Limit the trial logs a subset agents.
+  repeated determined.log.v1.LogLevel level = 9;
+  // Limit the trial logs a subset of output streams.
+  repeated string stdtype = 10;
+  // Limit the trial logs a subset of sources.
```
Comment (dzhu) [inline proto/src/determined/api/v1/trial.proto:39]:
```suggestion
  // Limit the trial logs to a subset of agents.
  repeated determined.log.v1.LogLevel level = 9;
  // Limit the trial logs to a subset of output streams.
  repeated string stdtype = 10;
  // Limit the trial logs to a subset of sources.
```

File: proto/src/determined/api/v1/trial.proto [hunk 6]
```diff
@@ -20,7 +26,24 @@ message TrialLogsRequest {
...
+  repeated string source = 11;
+  // Limit the trial logs where the timestamp is before a given time.
+  google.protobuf.Timestamp timestamp_before = 12;
+  // Limit the trial logs where the timestamp is after a given time.
```
Comment (dzhu) [inline proto/src/determined/api/v1/trial.proto:43]:
```suggestion
  // Limit the trial logs to ones with a timestamp before a given time.
  google.protobuf.Timestamp timestamp_before = 12;
  // Limit the trial logs to ones with a timestamp after a given time.
```
or something.

**Review by dzhu — APPROVED** [review]

File: master/internal/api/logs.go [hunk 7]
```diff
@@ -6,121 +6,171 @@ import (
...
+		req:            req,
+		fetcher:        fetcher,
+		process:        process,
+		terminateCheck: terminateCheck,
+		batchWaitTime:  batchWaitTime,
+	}
+}
+
+type (
+	tick struct{}
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/api/logs.go:75]: I like that.

File: master/internal/api/logs.go [hunk 8]
```diff
@@ -6,121 +6,171 @@ import (
...
+)
+
+// Receive implements the actor.Actor interface.
+func (l *LogStoreProcessor) Receive(ctx *actor.Context) error {
```
Comment (stoksc) [inline master/internal/api/logs.go:79]:
```suggestion

// Receive implements the actor.Actor interface.
func (l *LogStoreProcessor) Receive(ctx *actor.Context) error {
	type tick struct{}
```

**Pull request merged**